doc/specifications/api/lazy_properties.md

```diff
@@ -0,0 +1,214 @@
+# Lazy object properties
+
+
+As PHP API in eZ Platform is often used to populate templates via generic controllers, there is often a
+need to provide possibility to easily get additional info relevant directly to the domain objects involved.
+
+Prior to this feature this is only possible by:
+- Write own full fledge controller losing out on many content features by the system
+- Overriding the view controller where you can do almost anything you want, but you'll need to then maintain it yourself
+  and resort to PHP for data which in some cases are pretty basic need for the template.
+- Resort to query type which is limited to one set of objects, and mainly suitable to load for instance child objects.
+- Use of third party solutions which in varying degree diverges from the official supported API.
+
+While relevant for all domain objects, the most pressing need is for the eZ content model to get:
+- content info -> content type
+- content info -> section
+- content info -> main location _(just meta info\*)_
+- content info -> locations _(just meta info\*)_
```

Comment:
This (and main Location above) would have to respect https://github.com/ezsystems/ezpublish-kernel/blob/a1045266514a6984e9d05ec8547d2214e1d73410/eZ/Publish/API/Repository/LocationService.php#L68

AFAICS, that means using API in the generator. But if that is done, what happens with layered Repository implementations?

Reply:
Good point, yeah some of this should wait until further de-coupling of repository is done I guess. As otherwise way to much logic starts to move into domain mappers here. Or what did you have in mind in regards to `layered Repository implementations`?

The list here is just examples btw, could be that some of them rather make sense on content for instance. Either way intention is to do them one by one as followup with proper discussion on the use case involving PM _(most likely @bdunogier being the one to followup on that)_.

Reply:
I meant about Repository API implementations like SignalSlot, SiteAccessAware, at one point we also discussed having Cache, BusinessLogic and Permission implementations. Not sure what is the current status on those.

In the example of loading Content's Locations, which depends on permissions -- which implementation should be used in a generator?

BTW, loading Sections also has permission checks.

Reply:
> Cache, BusinessLogic and Permission implementations

currently not on the table I think, and if they are again in the future they will indeed affect how we inject these properties.

Reply:
It also applies to SiteAccessAware. For example, if I get SiteAccessAware implementation to load the correct translations, I want that to be respected for lazy-loaded properties. And since object mapping happens in the core layer, SiteAccessAware Repository would need to be propagated down to it.

Reply:
I know, it will inject the languages down to the api call (unless user specified something else). And Service+domainMapper will need to pass on those languages to the misc lazy properties builders as well.


Note: This does not solve/cover things like "Site awareness", like changing logic to filter things by hidden state, trash, .. Outside the scope of this PR, but one option would be to introduce new properties, for instance `->avaiableRelations` where _available_ or some other term implies logic filtering on what is accessible to the user, such property can and should be lazy.

Reply:
I wonder if the added complexity would justify itself. We already discussed this, but maybe it's time again to pitch for a dedicated layer, instead of making the Repo more complex. I would rather see it getting decoupled and simpler, and have the missing stuff built on top of it.

IMHO the biggest win of a dedicated layer is that you can do dedicated things there ;) For what it's worth, Site API is now ~1.5 years old, we use it on all our new projects, and I would still do it the same way.

Reply:
I'm aware, it's not up to me anymore and I'm not completely against. But the downside is that concept gets new meaning in  the other layer, so a property / method / concept has a two or more different meanings, depending on which layer you are in.

This change does not block such a path in the future, but for now the builtin in repo is injecting based, and value objects purely vary on the injected arguments by user or by the system (with SiteAccessAware).

Reply:
* Including possibility of eZ taking over SiteAPI and adapt it for official namespaces and adjust it for being able to support it.  Either way, it's in our interest to align with some/most of the efforts done in SiteAPI as it's tried and true already, and contains good concepts for making developers more productive on Platform.

Reply:
> But the downside is that concept gets new meaning in the other layer, so a property / method / concept has two or more different meanings, depending on which layer you are in.

In practice, Site API is used for everything on the read side of things, and you only need to use Public API for writing. That is usually smaller subset of codebase and also more or less separated from the rest. Even when it's not, with proper IDE type hinting and variable naming it's not a problem. At least nobody complained about the dichotomy so far, and developers assure me it has never been an issue. Ideally different layers would be reflected in naming, but when I try to think about it, it does not seem like an improvement.

But I can appreciate your concern. I would still make an argument that it is worth it. Having a dedicated layer that's declared as an "end" layer, something to be consumed and thrown away, enables you to do things you can't do with Repository API. We shouldn't try to turn Repository API into Site API, it was intentionally designed as verbose and general purpose, and I can't help but think that trying to do so will produce a poor result on both sides.

IMHO, having a good Site API is too important to be done with compromises, and that's reason enough for a separate layer.

> it's in our interest to align with some/most of the efforts done in SiteAPI

Aside from naming, what specifically would you avoid?

Two things that come to me are:

1. FilterService, ideally it would lean on filtering in Public API, which is missing ATM but I guess should happen at some point.
2. Relations -- I would really like to see proper implementation of named relations, instead of doing tricks with fields and tags. Also Location-based relations. Not sure if there are any plans for these, if not with new admin we might actually go for it.

Reply:
BTW, would it make sense to move this to https://discuss.ezplatform.com?

Reply:
We probably should, some time :)

Reply:
> so API/Site re -thinking api completely for Site usage.

Yes, I though that was implied :)

> btw, what did you imply here?

I discussed this with @hknezevic at some length, one of the reasons we didn't go for it so far is because it would need user interface in the Legacy admin UI. With new eZ Platform Admin UI v2, it might be worth our while.

Reply:
So location relations is on our radar, we'll probably need to add it over the next year.
But unsure about named relations, what did you have in mind there?

Reply:
It's somewhat similar to what was specified for refactoring some years ago, I don't remember all the details right now. Basically relations would have arbitrary string identifier for a type (name) and would be elevated to a first-class citizen. Right now we use field identifier to carry that meaning, but IMHO it's awkward. With this you wouldn't even have to edit the Content in order to change relations (they would apply on the published version, so could still be versioned).

Come to think of it, tags come into similar category, so potentially the same could be done with them as well. Just talked with @emodric and apparently Content-level tags were requested several times.

Reply:
> I don't remember "named", and don't immediately see the benefit of it.

I don't think we called it that, but there was some kind of spec/discussion for relations of arbitrary type. Benefits would be not depending on ContentType and ability to read and edit relations without having to read/edit Content.

> Asset relations exists in the roadmap

Is there more info about asset relations somewhere?